app/src/main/java/org/mozilla/fenix/SearchWidgetProvider.kt

```diff
@@ -0,0 +1,131 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.fenix
+
+import android.app.PendingIntent
+import android.appwidget.AppWidgetManager
+import android.appwidget.AppWidgetManager.OPTION_APPWIDGET_MIN_WIDTH
+import android.appwidget.AppWidgetProvider
+import android.content.Context
+import android.content.Intent
+import android.os.Bundle
+import android.widget.RemoteViews
+
+class SearchWidgetProvider : AppWidgetProvider() {
+
+    override fun onUpdate(context: Context, appWidgetManager: AppWidgetManager, appWidgetIds: IntArray) {
+        val textSearchIntent = createTextSearchIntent(context)
+        val voiceSearchIntent = createVoiceSearchIntent(context)
+
+        appWidgetIds.forEach { appWidgetId ->
+            val currentWidth = appWidgetManager.getAppWidgetOptions(appWidgetId).getInt(OPTION_APPWIDGET_MIN_WIDTH)
+            val layoutSize = getLayoutSize(currentWidth)
+            val layout = getLayout(layoutSize)
+            val text = getText(layoutSize, context)
+
+            val views = createRemoteViews(context, layout, textSearchIntent, voiceSearchIntent, text)
+            appWidgetManager.updateAppWidget(appWidgetId, views)
+        }
+    }
+
+    override fun onAppWidgetOptionsChanged(
+        context: Context,
+        appWidgetManager: AppWidgetManager,
+        appWidgetId: Int,
+        newOptions: Bundle?
+    ) {
+        val textSearchIntent = createTextSearchIntent(context)
+        val voiceSearchIntent = createVoiceSearchIntent(context)
+
+        val currentWidth = appWidgetManager.getAppWidgetOptions(appWidgetId).getInt(OPTION_APPWIDGET_MIN_WIDTH)
+        val layoutSize = getLayoutSize(currentWidth)
+        val layout = getLayout(layoutSize)
+        val text = getText(layoutSize, context)
+
+        val views = createRemoteViews(context, layout, textSearchIntent, voiceSearchIntent, text)
+        appWidgetManager.updateAppWidget(appWidgetId, views)
+    }
+
+    private fun getLayoutSize(dp: Int) = when {
+        dp >= DP_LARGE -> SearchWidgetProviderSize.LARGE
+        dp >= DP_MEDIUM -> SearchWidgetProviderSize.MEDIUM
+        dp >= DP_SMALL -> SearchWidgetProviderSize.SMALL
+        else -> SearchWidgetProviderSize.EXTRA_SMALL
+    }
+
+    private fun getLayout(size: SearchWidgetProviderSize) = when (size) {
+        SearchWidgetProviderSize.LARGE,
+            SearchWidgetProviderSize.MEDIUM -> R.layout.search_widget_large
+        SearchWidgetProviderSize.SMALL -> R.layout.search_widget_small
+        SearchWidgetProviderSize.EXTRA_SMALL -> R.layout.search_widget_extra_small
+    }
+
+    private fun getText(layout: SearchWidgetProviderSize, context: Context) = when (layout) {
+        SearchWidgetProviderSize.MEDIUM -> context.getString(R.string.search_widget_text_short)
```

Comment:
Is it intended that a small size doesn't have any text? It seems like on my galaxy s9 there's definitely room for there to be text on the size a step up from the smallest (just the icon)

<img width="350" alt="image" src="https://user-images.githubusercontent.com/4400286/61655077-ab32e780-ac72-11e9-9aa4-f0c769c37909.png">

Same goes for the medium widget which seems to have room for more text: 

<img width="350" alt="image" src="https://user-images.githubusercontent.com/4400286/61655149-cd2c6a00-ac72-11e9-964f-c66256d3eda9.png">

I did notice that Firefox has the same behavior, so maybe this isn't possible to fix (device specific handling?).

Reply:
Yeah, the small size isn't supposed to have any text (which is why I think Verdi was hoping we can set specific dp sizes instead of expanding the whole white space to fill in the cells).